### PR TITLE
reorganize a bit; add mocking

### PR DESCRIPTION
**Pull request by douglasdavis**

### CODE REVIEW

**Comment by douglasdavis** [comment]
This PR does two things:

1. makes it such that `repartition_layer` returns a low level graph instance instead of a new `Array` collection. `Array.repartition` gets a few more lines that handles instantiating a new collection from the new layer after making the layer `MaterializedLayer`.
2. add mocking logic to `optimize.py` so that we just return the same array (a collection's typetracer array representation is of course invariant under repartitioning)

File: src/dask_awkward/lib/core.py [hunk 1]
```diff
@@ -557,6 +557,10 @@ def reset_meta(self) -> None:
 
     def repartition(self, npartitions=None, divisions=None, rows_per_partition=None):
         from dask_awkward.lib.structure import repartition_layer
+        from dask.highlevelgraph import MaterializedLayer
+
+        token = tokenize(self, npartitions, divisions, rows_per_partition)
```
Comment (martindurant) [inline src/dask_awkward/lib/core.py]: (I would argue that any mix of npartitions/rows-per/divisions that ends up with the same divisions a few lines down amounts to exactly the same thing and should have the same token)

File: src/dask_awkward/lib/core.py [hunk 2]
```diff
@@ -568,7 +572,20 @@ def repartition(self, npartitions=None, divisions=None, rows_per_partition=None)
         if rows_per_partition:
             divisions = list(range(0, nrows, rows_per_partition))
             divisions.append(nrows)
-        return repartition_layer(self, divisions)
+
+        new_layer = repartition_layer(self, key, divisions)
+        new_layer = MaterializedLayer(new_layer)
+        new_layer._opt_mock_materialized = True
```
Comment (martindurant) [inline src/dask_awkward/lib/core.py]: Survives pickle?

File: src/dask_awkward/lib/optimize.py [hunk 3]
```diff
@@ -168,6 +168,10 @@ def _opt_touch_all_layer_names(dsk: HighLevelGraph) -> list[str]:
     # return _layers_with_annotation(dsk, "ak_touch_all")
 
 
+def _opt_mock_materialized_layer_names(dsk: HighLevelGraph) -> list[str]:
+    return [n for n, v in dsk.layers.items() if hasattr(v, "_opt_mock_materialized")]
```
Comment (martindurant) [inline src/dask_awkward/lib/optimize.py]: Comment: materializedLayers which **we made**

File: src/dask_awkward/lib/optimize.py [hunk 4]
```diff
@@ -197,6 +201,21 @@ def _mock_output(layer):
...
+
+    assert isinstance(args[0], ak.Array)
+    return args[0]
+
+
+def _mock_materialized(layer):
+    """Update a layer to do nothing."""
+    new_layer = copy.deepcopy(layer)
+    for k in iter(new_layer.keys()):
+        new_layer.mapping[k] = (_do_nothing_return_array,) + new_layer[k][1:]
```
Comment (martindurant) [inline src/dask_awkward/lib/optimize.py]: I don't follow, how does this get around multiple partitions?

**Review by martindurant — COMMENTED** [review]
Does this have the effect of squashing column opt failures?

I assume the same mock applies to slice or any other similar layer.

I don't mind the moving things around at all, if you think it makes sense this way.

File: src/dask_awkward/lib/optimize.py (hunk 4 above)
Comment (douglasdavis) [reply src/dask_awkward/lib/optimize.py]: Yea I'm digging into this more now; I kind of assumed `len(new_layer.keys())` would be 1 due to the partition culling that happens in the optimization, but that doesn't appear to be the case. I need to look at the code that does the culling for the optimization

File: src/dask_awkward/lib/core.py (hunk 2 above)
Comment (douglasdavis) [reply src/dask_awkward/lib/core.py]: I'm not sure, we've do this elsewhere with `_opt_touch_all` which was the workaround introduced to make sure `dak.zip` would work.

**Comment by douglasdavis** [comment]
The warnings have gone away with
```
pytest tests/test_structure.py -k "repartition"
```
but with more digging around I'm unconvinced the changes I added here are solid yet

File: src/dask_awkward/lib/core.py (hunk 1 above)
Comment (douglasdavis) [reply src/dask_awkward/lib/core.py]: makes sense; just changed this

**Comment by douglasdavis** [comment]
OK nevermind, _I_ introduced warnings and then removed them! So now I'm not sure if we're accomplishing anything yet. Do you have some code that triggers the optimization failure?

**Pull request merged**